Fix incorrect typings

777PolarFox777 opened the pull request:
Fix #723
Fix effect dispatcher type. When declaring and effect like (payload, state) => {} the state type was infered as meta type.

semoal commented:
LGTM,
but probably could be interesting to update this section, with some additional notes, and a 100% wokring example
https://rematch.github.io/rematch/#/recipes/typescript

777PolarFox777 commented:
@semoal I've updated Typescript docs section, could you check it?

semoal commented:
@ShMcK Looks good to me, we can't go worst than we are now. I'm not sure if there's some way to implement testing on type inference, or something to avoid rollin' back.

777PolarFox777 commented:
@seamoal if this PR is fine, could you merge it and release a new version?

ShMcK commented:
Happy to merge and release. The typings fixings look great.

But as this is a breaking change in the typings, could we get a quick summary of how to update typings to the latest - both for the release and changelog.

777PolarFox777 commented:
> Happy to merge and release. The typings fixings look great.
> 
> But as this is a breaking change in the typings, could we get a quick summary of how to update typings to the latest - both for the release and changelog.

Well, you can specify dispatch type when declaring effects as `RematchDispatch<RootModel>`, not as `RematchDispatch<void>`. 

Also you don't need to extends your `RootModels` type from `Models`, this helps to show error, when trying to call non-existing model, like `dispatch.nonExistingModel.nonExistingAction()`. 

And I've fixed declaring effect with two arguments, `(payload, rootState) => {...}` was infered in types like payload and metadata which is incorrect. 

For more information  you can refer to TypeScript docs section, I've updated them according to types changes.

semoal commented:
@ShMcK letts :shipit:

ShMcK commented:
Released as @rematch/core @1.4.0

tangdw commented:
`examples\ts\hooks\src\App.tsx`
![error](https://user-images.githubusercontent.com/17575065/82143023-5bfa1b00-9873-11ea-945b-2fa7a7b874ad.png)

leslie-heinzen commented:
Changes here are broken by https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#intersections-reduced-by-discriminant-properties

rematch is currently not compatible with TS 3.9
